File: server/medicore.layer/MediCore.Service/Model/HospitalAppointmentModel.cs
using dna.core.service.Models;
using dna.core.service.Models.Abstract;
using MediCore.Data.Infrastructure;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Threading.Tasks;

namespace MediCore.Service.Model
{
    public class HospitalAppointmentModel : WriteHistoryBaseModel, IModelBase
    {
        

        public int Id { get; set; }

        [Required(ErrorMessage = "Required")]
        public int HospitalId { get; set; }

        [Required(ErrorMessage = "Required")]
        public int MedicalStaffId { get; set; }

        [Required(ErrorMessage = "Required")]
        public DateTime AppointmentDate { get; set; }

        public DateTime? AppointmentStarted { get; set; }
        //fill when appointment finished
        public DateTime? AppointmentFinished { get; set; }

        [Required(ErrorMessage = "Required")]
        public AppointmentStatus AppointmentStatus { get; set; }

        [Required(ErrorMessage = "Required")]
        public int QueueNumber { get; set; }


        //if appointment create the by the logged user then fill it
        public int? UserId { get; set; }

        //if not registered user then he/she use PhoneNumber        

        [StringLength(50, ErrorMessage = "StringLength")]
        public string PhoneNumber { get; set; }

        //if not registered user then he/she use PatientName   
        [StringLength(255, ErrorMessage = "StringLength")]
        public string PatientName { get; set; }


        //if not registered user fill on PatientProblems
        //if registered user, fill on HospitalAppointmentDetail
        [StringLength(300, ErrorMessage = "StringLength")]
        public string PatientProblems { get; set; }

        //if user cancelled, fill the reason
        //cancelled status is from AppointmentStatus
        [StringLength(300, ErrorMessage = "StringLength")]
        public string CancelledReason { get; set; }

        public virtual UserModel User { get; set; }

       
        public virtual HospitalModel Hospital { get; set; }

       
        public virtual MedicalStaffModel MedicalStaff { get; set; }

        //token from device for sending FCM, not saved in DB
        public string DeviceId { get; set; }

        public virtual ICollection<HospitalAppointmentDetailModel> AppointmentDetails { get; set; }

        // not mapped
        public int CurrentQueueNumber { get; set; }
    }
}

File: server/medicore.apps/src/model/hospitalAppointmentModel.ts
import { Type } from 'class-transformer';
import { IEntityBase } from './IEntityBase';
import { AppointmentStatus } from './enumModel';
import { UserModel } from './userModel';
import { HospitalModel } from './hospitalModel';
import { MedicalStaffModel } from './medicalStaffModel';
import { HospitalAppointmentDetailModel } from './hospitalAppointmentDetailModel';

export class HospitalAppointmentModel implements IEntityBase {
    id: number;
    hospitalId: number;
    medicalStaffId: number;
    appointmentDate: Date;
    appointmentStatus: AppointmentStatus;
    queueNumber: number;

    //if appointment create the by the logged user then fill it
    userId: number;

    //if not registered user then use their PhoneNumber
    phoneNumber: string;

    //if not registered user then he/she use PatientName   
    patientName: string;

    //if not registered user fill on PatientProblems
    //if registered user, fill on HospitalAppointmentDetail
    patientProblems: string;

    //if user cancelled, fill the reason
    //cancelled status is from AppointmentStatus
    cancelledReason: string;

    //region model
    @Type(() => UserModel)
    user: UserModel;
    @Type(() => HospitalModel)
    hospital: HospitalModel;
    @Type(() => MedicalStaffModel)
    medicalStaff: MedicalStaffModel;
    @Type(() => HospitalAppointmentDetailModel)
    appointmentDetails: Array<HospitalAppointmentDetailModel>;

    constructor() {
        this.id = 0;
        this.hospitalId = 0;
        this.medicalStaffId = 0;
        this.appointmentDate = null;
        this.appointmentStatus = null;
        this.queueNumber = 0;
        this.userId = 0;
        this.phoneNumber = null;
        this.patientName = null;
        this.patientProblems = null;
        this.cancelledReason = null;
        this.user = null;
        this.hospital = null;
        this.medicalStaff = null;
        this.appointmentDetails = null;
    }
}